File: raffia-jackson/src/main/java/com/codepoetics/raffia/jackson/JsonWriter.java
package com.codepoetics.raffia.jackson;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.visitors.Visitors;
import com.codepoetics.raffia.writers.BasketWriter;
import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonGenerator;

import java.io.IOException;
import java.io.OutputStream;
import java.io.StringWriter;
import java.io.Writer;
import java.math.BigDecimal;

public final class JsonWriter implements BasketWriter<JsonWriter> {

  private static final JsonFactory DEFAULT = new JsonFactory();

  public static String writeBasketAsString(Basket basket, JsonFactory jsonFactory) {
    StringWriter writer = new StringWriter();
    writeBasketTo(basket, writer, jsonFactory);
    return writer.toString();
  }

  public static String writeBasketAsString(Basket basket) {
    StringWriter writer = new StringWriter();
    writeBasketTo(basket, writer);
    return writer.toString();
  }

  public static JsonWriter writeBasketTo(Basket basket, OutputStream outputStream, JsonFactory jsonFactory) {
    return basket.visit(Visitors.writingTo(writingTo(jsonFactory, outputStream)));
  }

  public static JsonWriter writeBasketTo(Basket basket, Writer writer, JsonFactory jsonFactory) {
    return basket.visit(Visitors.writingTo(writingTo(jsonFactory, writer)));
  }

  public static JsonWriter writeBasketTo(Basket basket, OutputStream outputStream) {
    return writeBasketTo(basket, outputStream, DEFAULT);
  }

  public static JsonWriter writeBasketTo(Basket basket, Writer writer) {
    return writeBasketTo(basket, writer, DEFAULT);
  }

  public static JsonWriter writingTo(OutputStream outputStream) {
      return writingTo(DEFAULT, outputStream);
  }

  public static JsonWriter writingTo(Writer writer) {
    return writingTo(DEFAULT, writer);
  }

  public static JsonWriter writingTo(JsonFactory factory, OutputStream outputStream) {
    try {
      return writingTo(factory.createGenerator(outputStream));
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static JsonWriter writingTo(JsonFactory factory, Writer writer) {
    try {
      return writingTo(factory.createGenerator(writer));
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static JsonWriter writingTo(JsonGenerator generator) {
    return new JsonWriter(null, StructContext.NONE, generator);
  }

  private enum StructContext {
    OBJECT,
    ARRAY,
    NONE
  }

  private final JsonWriter parent;
  private final StructContext context;
  private final JsonGenerator generator;

  public JsonWriter(JsonWriter parent, StructContext context, JsonGenerator generator) {
    this.parent = parent;
    this.context = context;
    this.generator = generator;
  }

  @Override
  public JsonWriter beginObject() {
    try {
      generator.writeStartObject();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return new JsonWriter(this, StructContext.OBJECT, generator);
  }

  @Override
  public JsonWriter beginArray() {
    try {
      generator.writeStartArray();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return new JsonWriter(this, StructContext.ARRAY, generator);
  }

  @Override
  public JsonWriter end() {
    try {
      switch (context) {
        case ARRAY: {
          generator.writeEndArray();
          return flushOnCompletion(parent);
        }
        case OBJECT: {
          generator.writeEndObject();
          return flushOnCompletion(parent);
        }
        default:
          throw new IllegalStateException("end called without matching startArray or startObject");
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  private JsonWriter flushOnCompletion(JsonWriter writer) {
    if (writer.context.equals(StructContext.NONE)) {
      try {
        generator.flush();
      } catch (IOException e) {
        throw new RuntimeException(e);
      }
    }
    return writer;
  }

  @Override
  public JsonWriter key(String key) {
    try {
      generator.writeFieldName(key);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return this;
  }

  @Override
  public JsonWriter add(String value) {
    try {
      generator.writeString(value);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return flushOnCompletion(this);
  }

  @Override
  public JsonWriter add(BigDecimal value) {
    try {
      generator.writeNumber(value);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return flushOnCompletion(this);
  }

  @Override
  public JsonWriter add(boolean value) {
    try {
      generator.writeBoolean(value);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return flushOnCompletion(this);
  }

  @Override
  public JsonWriter addNull() {
    try {
      generator.writeNull();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return flushOnCompletion(this);
  }
}

File: raffia-jackson/src/main/java/com/codepoetics/raffia/jackson/JsonWriter.kt
package com.codepoetics.raffia.jackson

import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.writers.BasketWriter
import com.codepoetics.raffia.writers.writeTo
import com.fasterxml.jackson.core.JsonFactory
import com.fasterxml.jackson.core.JsonGenerator

import java.io.IOException
import java.io.OutputStream
import java.io.StringWriter
import java.io.Writer
import java.math.BigDecimal

class JsonWriter private constructor(private val generator: JsonGenerator) : BasketWriter<JsonWriter> {

    private enum class StructContext {
        OBJECT,
        ARRAY,
        NONE
    }

    private val contextStack = mutableListOf<StructContext>(StructContext.NONE)
    private var stackPtr = 0

    override fun beginObject(): JsonWriter = apply {
        generator.writeStartObject()

        stackPtr += 1
        if (stackPtr == contextStack.size) contextStack.add(StructContext.OBJECT) else contextStack[stackPtr] = StructContext.OBJECT
    }

    override fun beginArray(): JsonWriter = apply {
        generator.writeStartArray()

        stackPtr += 1
        if (stackPtr == contextStack.size) contextStack.add(StructContext.ARRAY) else contextStack[stackPtr] = StructContext.ARRAY
    }

    private inline fun update(block: JsonWriter.() -> Unit) = apply {
        block()
        if (stackPtr == 0) generator.flush()
    }

    override fun end(): JsonWriter =
        when (contextStack[stackPtr]) {
            JsonWriter.StructContext.ARRAY -> update {
                generator.writeEndArray()
                stackPtr -= 1
            }
            JsonWriter.StructContext.OBJECT -> update {
                generator.writeEndObject()
                stackPtr -= 1
            }
            else -> throw IllegalStateException("exit called without matching startArray or startObject")
        }

    override fun key(key: String): JsonWriter = apply {
        generator.writeFieldName(key)
    }

    override fun add(value: String): JsonWriter = update {
        generator.writeString(value)
    }

    override fun add(value: BigDecimal): JsonWriter = update {
        generator.writeNumber(value)
    }

    override fun add(value: Boolean): JsonWriter = update {
        generator.writeBoolean(value)
    }

    override fun addNull(): JsonWriter = update {
        generator.writeNull()
    }

    companion object {

        private val DEFAULT = JsonFactory()

        @JvmStatic
        fun writeBasketAsString(basket: Basket, jsonFactory: JsonFactory): String {
            val writer = StringWriter()
            writeBasketTo(basket, writer, jsonFactory)
            return writer.toString()
        }

        @JvmStatic
        fun writeBasketAsString(basket: Basket): String {
            val writer = StringWriter()
            writeBasketTo(basket, writer)
            return writer.toString()
        }

        @JvmStatic
        @JvmOverloads fun writeBasketTo(basket: Basket, outputStream: OutputStream, jsonFactory: JsonFactory = DEFAULT): JsonWriter {
            return basket.writeTo(writingTo(jsonFactory, outputStream))
        }

        @JvmStatic
        @JvmOverloads fun writeBasketTo(basket: Basket, writer: Writer, jsonFactory: JsonFactory = DEFAULT): JsonWriter {
            return basket.writeTo(writingTo(jsonFactory, writer))
        }

        @JvmStatic
        fun writingTo(outputStream: OutputStream): JsonWriter {
            return writingTo(DEFAULT, outputStream)
        }

        @JvmStatic
        fun writingTo(writer: Writer): JsonWriter {
            return writingTo(DEFAULT, writer)
        }

        @JvmStatic
        fun writingTo(factory: JsonFactory, outputStream: OutputStream): JsonWriter {
            try {
                return writingTo(factory.createGenerator(outputStream))
            } catch (e: IOException) {
                throw RuntimeException(e)
            }

        }

        @JvmStatic
        fun writingTo(factory: JsonFactory, writer: Writer): JsonWriter {
            try {
                return writingTo(factory.createGenerator(writer))
            } catch (e: IOException) {
                throw RuntimeException(e)
            }

        }

        @JvmStatic
        fun writingTo(generator: JsonGenerator): JsonWriter {
            return JsonWriter(generator)
        }
    }
}
